lib/mrb/scripts/logger/flag.rb

```diff
@@ -0,0 +1,36 @@
+module Groonga
+  class Logger
+    class Flag
+      @@flags = {}
+      class << self
+        def set_flags(flag)
+          @@default_flags = flag
+          self
+        end
+
+        def format
+          flag_names = @@flags.collect do |index, candidate|
+            candidate.name if candidate.flag & @@default_flags > 0
+          end
+          flag_names.compact.join("|")
+        end
+      end
+
+      attr_reader :name, :flag
+      def initialize(name, flag)
+        @@flags[flag] = self
+        @name  = name
+        @flag = flag
+      end
+
+      NONE       = new(:none,       0)
+      TIME       = new(:time,       1)
+      TITLE      = new(:title,      2)
+      MESSAGE    = new(:message,    4)
+      LOCATION   = new(:location,   8)
+      PID        = new(:pid,        16)
+      PROCESS_ID = new(:process_id, 16)
+      THREAD_ID  = new(:thread_id,  32)
+    end
```

Comment:
Umm. This class isn't a good API.

How about the following?

```ruby
module Groonga
  class Logger
    class Flags
      VALUES = {
        time: 1,
        title: 2,
        message: 4,
        location: 8,
        process_id: 16,
        thread_id: 32,
      }

      attr_reader :value
      def initialize(value)
        case value
        when String, Symbol
          @value = VALUES[value.to_sym]
          if @value.nil?
            available_names = VALUES.keys.inspect
            message = "unknown flag name: #{value.inspect}: #{available_names}"
            raise ArgumentError, message
          end
        else
          @value = value
        end
      end

      def to_i
        @value
      end

      def to_s
        names = []
        VALUES.each do |name, value|
          names << name.to_s unless (value & @value) == 0
        end
        names.join("|")
      end

      def |(other)
        other = self.class.new(other) unless other.is_a?(self.class)
        new(@value | other.to_i)
      end

      def &(other)
        other = self.class.new(other) unless other.is_a?(self.class)
        new(@value & other.to_i)
      end

      NONE = new(0)
      TIME = new(:time)
      TITLE = new(:title)
      MESSAGE = new(:message)
      LOCATION = new(:location)
      PROCESS_ID = new(:process_id)
      PID = PROCESS_ID
      THREAD_ID = new(:thread_id)
    end
  end
end
```

```c
static mrb_value
logger_s_get_default_flags(mrb_state *mrb, mrb_value self)
{
  mrb_value mrb_flags_class =
    mrb_const_get(mrb, self, mrb_intern_lit(mrb, "Flag"));
  mrb_value mrb_flags = mrb_fixnum_value(grn_default_logger_get_flags());
  return mrb_funcall(mrb, mrb_flags_class, "new", 1, mrb_flags);
}
```

Reply:
It seems that it is reasonable API which I couldn't design.